db/write_thread.cc

```diff
@@ -656,7 +650,27 @@ void WriteThread::ExitAsBatchGroupLeader(WriteGroup& write_group,
   }
 
   if (enable_pipelined_write_) {
-    // Notify writers don't write to memtable to exit.
+    // We insert a dummy Writer right before our current write_group. This
+    // allows us to unlink our write_group without the risk that a subsequent
+    // writer becomes a new leader and might overtake us and add itself to the
+    // memtable-writer-list before we can do so. This ensures that writers are
+    // added to the memtable-writer-list in the exact same order in which they
+    // were in the newest_writer list.
+
+    // This must happen before completing the writers from our group to prevent
+    // a race where the owning thread of one of these writers can start a new
+    // write operation.
+    Writer dummy;
+    Writer* head = newest_writer_.load(std::memory_order_acquire);
```

Comment:
The start of this closely resembles the `!enable_pipelined_write_` branch below, which has some useful comments. I would add a comment to refer to it. (Not obvious how to cleanly combine the two pieces of code--without introducing an unnecessary allocation of dummy Writer on the stack.)

Reply:
I copied some comments here to explain what is being done, hoping this duplication is easier to maintain (because comment is close to code) than referring to another comment. Should we have a way of identifying comments, I may have used a reference.